// powerplatform-mcp/build/PowerPlatformClient.js
import { ConfidentialClientApplication } from '@azure/msal-node';
import axios from 'axios';
/**
 * Base client for PowerPlatform API access.
 * Handles authentication and generic HTTP requests.
 * Service classes should depend on this client via constructor injection.
 */
export class PowerPlatformClient {
    config;
    msalClient;
    accessToken = null;
    tokenExpirationTime = 0;
    constructor(config) {
        this.config = config;
        // Initialize MSAL client
        this.msalClient = new ConfidentialClientApplication({
            auth: {
                clientId: this.config.clientId,
                clientSecret: this.config.clientSecret,
                authority: `https://login.microsoftonline.com/${this.config.tenantId}`,
            }
        });
    }
    /**
     * Get the organization URL
     */
    get organizationUrl() {
        return this.config.organizationUrl;
    }
    /**
     * Get an access token for the PowerPlatform API
     */
    async getAccessToken() {
        const currentTime = Date.now();
        // If we have a token that isn't expired, return it
        if (this.accessToken && this.tokenExpirationTime > currentTime) {
            return this.accessToken;
        }
        try {
            // Get a new token
            const result = await this.msalClient.acquireTokenByClientCredential({
                scopes: [`${this.config.organizationUrl}/.default`],
            });
            if (!result || !result.accessToken) {
                throw new Error('Failed to acquire access token');
            }
            this.accessToken = result.accessToken;
            // Set expiration time (subtract 5 minutes to refresh early)
            if (result.expiresOn) {
                this.tokenExpirationTime = result.expiresOn.getTime() - (5 * 60 * 1000);
            }
            return this.accessToken;
        }
        catch (error) {
            console.error('Error acquiring access token:', error);
            throw new Error('Authentication failed');
        }
    }
    /**
     * Make an authenticated GET request to the PowerPlatform API
     * @param endpoint The API endpoint (relative to organization URL)
     */
    async get(endpoint) {
        try {
            const token = await this.getAccessToken();
            const response = await axios({
                method: 'GET',
                url: `${this.config.organizationUrl}/${endpoint}`,
                headers: {
                    'Authorization': `Bearer ${token}`,
                    'Accept': 'application/json',
                    'OData-MaxVersion': '4.0',
                    'OData-Version': '4.0'
                }
            });
            return response.data;
        }
        catch (error) {
            console.error('PowerPlatform API request failed:', error);
            throw new Error(`PowerPlatform API request failed: ${error}`);
        }
    }
    /**
     * Make an authenticated POST request to the PowerPlatform API
     * @param endpoint The API endpoint (relative to organization URL)
     * @param data The request body
     */
    async post(endpoint, data) {
        try {
            const token = await this.getAccessToken();
            const response = await axios({
                method: 'POST',
                url: `${this.config.organizationUrl}/${endpoint}`,
                headers: {
                    'Authorization': `Bearer ${token}`,
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                    'OData-MaxVersion': '4.0',
                    'OData-Version': '4.0'
                },
                data
            });
            return response.data;
        }
        catch (error) {
            console.error('PowerPlatform API POST request failed:', error);
            throw new Error(`PowerPlatform API POST request failed: ${error}`);
        }
    }
}
