// @mcp-consultant-tools/powerplatform/build/auth/index.js
/**
 * Authentication module for PowerPlatform MCP Server
 *
 * Supports two authentication modes:
 * 1. Service Principal (ConfidentialClientApplication) - when client_secret is provided
 * 2. Interactive User Auth (PublicClientApplication) - when no client_secret is provided
 */
import { ServicePrincipalAuth } from './service-principal-auth.js';
import { InteractiveAuth } from './interactive-auth.js';
/**
 * Create an appropriate auth provider based on configuration
 *
 * If clientSecret is provided → ServicePrincipalAuth (existing behavior)
 * If no clientSecret → InteractiveAuth (browser-based SSO)
 *
 * @param config - PowerPlatform authentication configuration
 * @returns Auth provider instance
 */
export function createAuthProvider(config) {
    if (config.clientSecret) {
        // Service Principal mode (existing behavior)
        return new ServicePrincipalAuth({
            organizationUrl: config.organizationUrl,
            clientId: config.clientId,
            clientSecret: config.clientSecret,
            tenantId: config.tenantId,
        });
    }
    // Interactive User Auth mode (new behavior)
    return new InteractiveAuth({
        organizationUrl: config.organizationUrl,
        clientId: config.clientId,
        tenantId: config.tenantId,
    });
}
export { ServicePrincipalAuth } from './service-principal-auth.js';
export { InteractiveAuth } from './interactive-auth.js';
export { TokenCache } from './token-cache.js';
